lib/gc/Transforms/BufferHoist.cpp

```diff
@@ -0,0 +1,375 @@
+          }
+        }
+        MemRefType memrefType = dyn_cast<MemRefType>(allocValue.getType());
+        bool isStaticShape = memrefType && memrefType.hasStaticShape();
+        if (!parallelOpsInCurBlock.empty() && isStaticShape) {
+          for (auto *use : allocValue.getUsers()) {
+            for (auto *operation : parallelOpsInCurBlock) {
+              if (operation->isAncestor(use)) {
+
+                // only support scf.forall for now
+                if (auto forallOp = dyn_cast<scf::ForallOp>(parentOp)) {
+                  SmallVector<Value> ubs = getValueOrCreateConstantIndexOp(
+                      builder, forallOp.getLoc(),
+                      forallOp.getMixedUpperBound());
+                  if (std::optional<int64_t> ubs0_int =
+                          getConstantIntValue(ubs[0])) {
+                    numThreads = ubs0_int.value();
+                    shouldHoist = true;
+                  }
+                  inductVar = forallOp.getInductionVar(0);
+                }
+              }
+            }
+          }
+        }
+      }
+
+      // Find the actual placement block and determine the start operation using
+      // an upper placement-block boundary. The idea is that placement block
+      // cannot be moved any further upwards than the given upper bound.
+      Block *placementBlock = findPlacementBlock(
+          state, state.computeUpperBound(dominatorBlock, dependencyBlock),
+          shouldHoist);
+      Operation *startOperation = BufferPlacementAllocs::getStartOperation(
+          allocValue, placementBlock, liveness);
+
+      // Move the alloc in front of the start operation.
+      Operation *allocOperation = allocValue.getDefiningOp();
+      if (shouldHoist) {
+        OpBuilder builder(allocOperation);
+        // 1) allocate larger buffer...
+        //   1.1) query the forall op to get the inductor var and the upperBound
+        //   1.2) create new memref type with larger size based on upperBound
+        //   1.3) create new allocOp with the new memref type
+        //   1.4) move the new allocOp in the correct insertpoint.
+        // 2) replace with new memref::subViewOp
+        //   2.1) create a new memref::subViewOp op based on the new allocOp
+        //   2.2) replace the original allocOp and its use with the new
+        //   memref::subViewOp
+        //   2.2) remove the original allocOp
+        MemRefType memrefType = dyn_cast<MemRefType>(allocValue.getType());
+        if (memrefType && memrefType.hasStaticShape()) {
+          Type elementType = memrefType.getElementType();
+          int64_t memrefTypeRank = memrefType.getRank();
+
+          ArrayRef<int64_t> originalShape = memrefType.getShape();
+          // Create a new shape with the desired larger size
+          SmallVector<int64_t, 4> newShape(originalShape.begin(),
+                                           originalShape.end());
+          newShape[0] *= numThreads;
```

Comment:
Seems you're trying to hoist to the upper level of parallel for?

Reply:
Good point, it should be hoisted to the top of outer-most parallel for, will fix it.

Reply:
I think we need a discussion about this, that is, which is the right position for thraed-local memory allocation, to leverage thread-local allocator, the hoist will be inside the parallel for. For thread-group memory allocation, I think we want to hoist them to the outer-most parallel for.

Reply:
As discussed offline, this PR is targeting the 2nd case (thread local allocator within thread-group) only as it's specific to our thread-group usage. While the 1st case mentioned above is already covered by MLIR's `buffer-loop-hoist` pass which is more general.

Reply:
It's addressed now.